temporal-test-server/src/main/java/io/temporal/internal/testservice/TestService.java

```diff
@@ -0,0 +1,118 @@
+import java.io.Closeable;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * In memory implementation of the Operator Service. To be used for testing purposes only.
+ *
+ * <p>Do not use directly, instead use {@link io.temporal.testing.TestWorkflowEnvironment}.
+ */
+public final class TestService extends TestServiceGrpc.TestServiceImplBase implements Closeable {
+
+  private final TestWorkflowStore workflowStore;
+
+  public TestService(TestWorkflowStore workflowStore) {
+    this.workflowStore = workflowStore;
+  }
+
+  @Override
+  public void lockTimeSkipping(
+      LockTimeSkippingRequest request, StreamObserver<LockTimeSkippingResponse> responseObserver) {
+    lockTimeSkipping(request.getCallerId());
+    responseObserver.onNext(LockTimeSkippingResponse.newBuilder().build());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void unlockTimeSkipping(
+      UnlockTimeSkippingRequest request,
+      StreamObserver<UnlockTimeSkippingResponse> responseObserver) {
+    unlockTimeSkipping(request.getCallerId());
+    responseObserver.onNext(UnlockTimeSkippingResponse.newBuilder().build());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void skipTime(SkipTimeRequest request, StreamObserver<SkipTimeResponse> responseObserver) {
+    sleep(ProtobufTimeUtils.toJavaDuration(request.getDuration()));
+    responseObserver.onNext(SkipTimeResponse.newBuilder().build());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void getCurrentTime(
+      Empty request, StreamObserver<GetCurrentTimeResponse> responseObserver) {
+    Timestamp timestamp = workflowStore.currentTime();
+    responseObserver.onNext(GetCurrentTimeResponse.newBuilder().setTime(timestamp).build());
+    responseObserver.onCompleted();
+  }
+
+  /**
+   * Disables time skipping. To enable back call {@link #unlockTimeSkipping(String)}. These calls
+   * are counted, so calling unlock does not guarantee that time is going to be skipped immediately
+   * as another lock can be holding it.
+   */
+  public void lockTimeSkipping(String caller) {
+    workflowStore.getTimer().lockTimeSkipping(caller);
+  }
+
+  public void unlockTimeSkipping(String caller) {
+    workflowStore.getTimer().unlockTimeSkipping(caller);
+  }
+
+  /**
+   * Blocks calling thread until internal clock doesn't pass the current + duration time. Might not
+   * block at all due to time skipping.
+   */
+  public void sleep(Duration duration) {
+    CompletableFuture<Void> result = new CompletableFuture<>();
+    workflowStore
+        .getTimer()
+        .schedule(
+            duration,
+            () -> {
+              workflowStore.getTimer().lockTimeSkipping("TestWorkflowService sleep");
+              result.complete(null);
+            },
+            "workflow sleep");
+    workflowStore.getTimer().unlockTimeSkipping("TestWorkflowService sleep");
```

Comment:
The only situation when we need to use sleep is when the time skipping is locked. Otherwise time fast forwards without limits, no sleeps is needed.
What happens here, the timer gets scheduled. After that, the time gets unlocked. After that time gets locked by the fired timer that we just scheduled.